File: library-management/src/main/java/com/library/management/service/StatisticsService.java
package com.library.management.service;

import com.library.management.dto.statistics.LateRateDTO;
import com.library.management.dto.statistics.MonthlyBorrowingDTO;
import com.library.management.dto.statistics.TopAuthorDTO;

import java.util.List;

/**
 * Business operations for statistics/reporting.
 */
public interface StatisticsService {

    List<MonthlyBorrowingDTO> getBorrowingsPerMonth();

    List<TopAuthorDTO> getTopAuthors();

    LateRateDTO getLateRate();
}

File: library-frontend/src/services/statisticsService.js
import axiosInstance from './axiosInstance';

// Mirrors com.library.management.controller.StatisticsController
// Base path: /api/statistics
const statisticsService = {
  // Returns MonthlyBorrowingDTO[]: { month: "YYYY-MM", numberOfBorrowings }
  getBorrowingsPerMonth: () =>
    axiosInstance
      .get('/statistics/borrowings-per-month')
      .then((res) => res.data),

  // Returns TopAuthorDTO[]: { authorId, authorName, numberOfBorrowings }
  getTopAuthors: () =>
    axiosInstance.get('/statistics/top-authors').then((res) => res.data),

  // Returns LateRateDTO: { totalBorrowings, lateBorrowings, lateRate }
  getLateRate: () =>
    axiosInstance.get('/statistics/late-rate').then((res) => res.data),
};

export default statisticsService;
